packages/@uppy/core/src/Uppy.js

```diff
@@ -584,68 +646,50 @@ class Uppy {
   addFiles (fileDescriptors) {
     this.#assertNewUploadAllowed()
 
-    // create a copy of the files object only once
-    const files = { ...this.getState().files }
-    const newFiles = []
-    const errors = []
-    for (let i = 0; i < fileDescriptors.length; i++) {
-      try {
-        let newFile = this.#checkAndCreateFileStateObject(files, fileDescriptors[i])
-        // Users are asked to re-select recovered files without data,
-        // and to keep the progress, meta and everthing else, we only replace said data
-        if (files[newFile.id] && files[newFile.id].isGhost) {
-          newFile = {
-            ...files[newFile.id],
-            data: fileDescriptors[i].data,
-            isGhost: false,
-          }
-          this.log(`Replaced blob in a ghost file: ${newFile.name}, ${newFile.id}`)
-        }
-        files[newFile.id] = newFile
-        newFiles.push(newFile)
-      } catch (err) {
-        if (!err.isRestriction) {
-          errors.push(err)
-        }
-      }
-    }
+    const { nextFilesState, validFilesToAdd, errors } = this.#checkAndUpdateFileState(fileDescriptors)
 
-    this.setState({ files })
+    const restrictionErrors = errors.filter((error) => error.isRestriction)
+    this.#informAndEmit(restrictionErrors)
 
-    newFiles.forEach((newFile) => {
-      this.emit('file-added', newFile)
+    const nonRestrictionErrors = errors.filter((error) => !error.isRestriction)
+
+    // todo shouldn't we only setState if we are not throwing an error? (see below)
+    this.setState({ files: nextFilesState })
```

Comment:
🤔

Reply:
yes, I mean further down in the function we throw an error, e.g. `throw new AggregateError` etc.

If we call `setState({ files: nextFilesState })`, but then later throw an error, then the state has already been set. so i'm thinking we might end up in an invalid state where we have files set, even though the `addFiles` operation threw an error. to me that sounds like a broken implementation. Note that this is how the code already works, I just thought it might make sense to change this.

Reply:
Feel free to move it down if it makes more sense in the current implementation.

Reply:
yes i was just afraid to break something if i changed the behavior, but i guess we'll find out 😮